### ISSUE TITLE
gather never returns when a worker dies

### ISSUE THREAD

**Issue by farquet** [dask/distributed#1110]
Hello,

I run into an unexpected behavior using dask.
In my setup, it can happen that a worker gets suddenly killed. One would expect that a blocking distributed.client.gather() on the future of a task from that worker would return an error or throw an exception. But instead it gets stuck forever waiting for the dead worker to give back the results.

Is that a bug or should this be handled by the user ? If it's not a bug, what is the proper to detect the dead worker and resubmit the failed experiment ?

Thanks !

The system is stuck with these last debug statements : 
```
distributed.client - DEBUG - Submit <Runner object at 0x2aab2c910350>(...), <Runner object at 0x2aab2c910350>-1db47428ed100623e45496e48a87dede
distributed.client - DEBUG - Waiting on futures to clear before gather
```

**Comment by mrocklin** [comment]
It's not waiting on the worker, it's waiting on the computation to finish.  The scheduler will redirect the work to other workers as necessary.  My guess is that in your situation you have no other valid workers, so the task is never completed.

**Comment by farquet** [comment]
Alright, thanks for your answer! Indeed, all workers got killed. But when if a spawn new workers (using the same scheduler-file), I see the keep alive messages being exchanged with the scheduler, but the computation isn't scheduled on the new workers.

**Comment by mrocklin** [comment]
That surprises me.  Can you verify that your scheduler sees the workers.  This should be evident by printing out the client object and verifying that it sees workers.

```python
In [4]: client
Out[4]: <Client: scheduler='tcp://127.0.0.1:45400' processes=8 cores=8>
```

You then might also look at the futures to verify that they are pending

```python
In [7]: future
Out[7]: <Future: status: pending, key: sleep-bbb0c232ca9ea16bc9d057719b511727>
```

**Comment by farquet** [comment]
Indeed, this is exactly the state I am in :

```
Future : <Future: status: pending, key: <Runner object at 0x2aab2d9b9e10>-5415ba821e575e88d4400f0218f70c43>
client : <Client: scheduler=u'tcp://10.213.5.131:8786' processes=1 cores=32>
```

I reached that state by having N workers before they all got killed. I got down to 0. New tasks have been submitted and show a 'pending' status, as expected. A new worker is added, but the future stays in pending mode forever (the log that you see above).

**Comment by farquet** [comment]
Alright, so it appears that we were binding the task submission to specific workers... Then obviously, the other workers couldn't claim them. So, I am closing this issue.
However, that would have been nice to have the future tell you for which resources it is waiting.

Thanks for your help
